### PR TITLE
Add capability to disable specific alert rules Loki

### PR DESCRIPTION
**Pull request by Daniel-Vaz**
**What this PR does / why we need it**:

Currently using the Loki Chart we can only either enable\disable ALL alert rules. 
For specific environments and use-cases sometimes not all alert Rules are useful to have enabled.

With this PR change, we can cleanly and through the Chart values disable specific Alerts.

**Special notes for your reviewer**:

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] `values.yaml` comments added

Extra Note:

This is my first time doing a PR for this project. I forked the main branch and implemented these changes. Did not change the Chart.yaml version neither the Changelog, but  if needed (or any other missing action) just request me and I will do it.

Thank you in advance for all your awesome work.

### CODE REVIEW

**Review by dannykopping — COMMENTED** [review]
Hey @Daniel-Vaz 

I appreciate what you're trying to do but I think we need a more generic solution. Can we take a step back and talk about the problem you're trying to solve? Why do you want to disable these specific alerts?

**Comment by Daniel-Vaz** [comment]
Hi @dannykopping , thank you for the feedback !

>  Can we take a step back and talk about the problem you're trying to solve? Why do you want to disable these specific alerts?

So currently I'm a consumer of this Loki Chart, and since the clusters where I have Loki deployed I also have the Prom Operator, I want to enable the Chart suggested Scrapes\Rules\Alerts. 

The Problem I'm facing, is that in some of my clusters indeed we have some "issues" in the overall architecture that result on Loki reporting quite frequently some `LokiRequestLatency`. 

We could just silence them. 
Or even we could just disable all the Chart default alerts, and copy them on our side. 

But I believe that giving the flexibility to the Chart consumer to pick and chose which alerts they want to have enabled or not would be a nice feature.

For example, on the [Kube Prometheus Stack Chart they support this disabling specific alerts capability](https://github.com/prometheus-community/helm-charts/commit/51351ed2db55a180105598ac83255afde0a12765#diff-393dbf14eb23b32a190896847ef94be5563f651ba7cc85f9f8cf2254e6f15dbe). 

Do you have any other suggestion on how this capability to be implemented ?

**Review by dannykopping — APPROVED** [review]
Alright, that sounds like a good enough reason to me - and good to see some prior art where this is done the same way. Thanks for that

LGTM! I'll just request a PR from a colleague who knows Helm before we merge.

**Comment by shantanualsi** [comment]
@Daniel-Vaz have you tried testing when all alerts are set to false and the rules block under `loki_alerts` is empty?

**Comment by Daniel-Vaz** [comment]
> LGTM! I'll just request a PR from a colleague who knows Helm before we merge.

Thank you very much @dannykopping !

> @Daniel-Vaz have you tried testing when all alerts are set to false and the rules block under `loki_alerts` is empty?

@shantanualsi  Indeed in that situation the object will fail to be applied because it as no rules under the groups section. 
I assumed that it's not likely that people will disable all alerts one by one. In those situations they will just set the `monitoring.rules.alerting` value to false. 

Either way I think that if we find a way to address this corner-case it would indeed increase the reliability of the Chart. 
I'm just a bit unsure how could it be implemented :/ any ideas ?

**Comment by Daniel-Vaz** [comment]
> in our case, you can technically have `enabled: true` and `LokiCanaryLatency: true` and the chart will still break

The way I implemented it is not the case I believe. Since for the `LokiCanaryLatency` the If condition encapsulates the whole `loki_canaries_alerts` rules group section. 

```yaml
{{- if not (.Values.monitoring.rules.disabled.LokiCanaryLatency | default false) }}
  - name: "loki_canaries_alerts"
    rules:
      - alert: "LokiCanaryLatency"
        annotations:
          message: |
            {{`{{`}} $labels.job {{`}}`}} is experiencing {{`{{`}} printf "%.2f" $value {{`}}`}}s 99th percentile latency.
        expr: |
          histogram_quantile(0.99, sum(rate(loki_canary_response_latency_seconds_bucket[5m])) by (le, namespace, job)) > 5
        for: "15m"
        labels:
          severity: "warning"
{{- if .Values.monitoring.rules.additionalRuleLabels }}
{{ toYaml .Values.monitoring.rules.additionalRuleLabels | indent 10 }}
{{- end }}
{{- end }}
```

I did this because that rules group section had only that single alert. If more are added in he future, then that needs to be updated indeed. 

---

> The kube-prometheus-stack you referenced in one of the comments above has a python script under hack folder that creates yaml files.. do you know if it addresses this particular case?

I believe that hack script is used in order to have a sync from a external place where alerts are defined. They dont write their own alerts definitions inside the Kube-Prometheus-Stack Chart, they instead just use those script to sync the rules from a external place and format them accordingly. 

The usage for those scripts is written [here](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack/hack).


---

> I'm also thinking that rules section now has enabled which is a boolean and disabled which is a map of <rules>: true. This can be confusing to other users.

I kind of agree with you on this one. 
But at the same time the name of those values is quite self explanatory no ?
- If you want to enable alerting then `monitoring.rules.alerting: true`
- If you need to disable a specific rule from the whole group, then : `monitoring.rules.disabled: {}`

We could also move the `disabled` to be nested under `alerting`.
Or even under `alerting` instead of just supporting <true/false> values, we can change the structure, and there specify the list of rules names, to be enabled\disabled, for example:

```yaml
monitoring:
  rules:
    alerting:
      LokiTooManyCompactorsRunning: true
      LokiRequestLatency: false
      
# By default all values should be true
```

**Comment by shantanualsi** [comment]
Sorry for the delay here. 
> Since for the LokiCanaryLatency the If condition encapsulates the whole loki_canaries_alerts rules group section.

Apologies, I wasn't clear enough here.. in case if all rules under `loki_alerts` are disabled and rules under `loki_canaries_alerts` are enabled. 

Can you please add some documentation on the values.yaml for the edge case mentioning to set `monitoring.rules.alerting`  to false explicitly if all rules are disabled? Thanks for your patience! :)

**Comment by Daniel-Vaz** [comment]
@shantanualsi  Thank you :D !

Updated the values.yaml comments, whenever you want feel free to have a look.

**Comment by JStickler** [comment]
Also, please note that when you update the Helm charts, you have to bump the version of the Helm Chart.  This is mentioned in the PR checklist.

For Helm chart changes bump the Helm chart version in production/helm/loki/Chart.yaml and update production/helm/loki/CHANGELOG.md and production/helm/loki/README.md. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)

**Comment by shantanualsi** [comment]
@Daniel-Vaz can you also please increment the helm version and add a changelog as suggested by @JStickler

**Comment by Daniel-Vaz** [comment]
@JStickler @shantanualsi 

Done, Thank you for the warning and sorry for missing that part previously.

File: production/helm/loki/values.yaml
Comment (JStickler) [inline production/helm/loki/values.yaml:566]:
```suggestion
    # -- If you disable all the alerts and keep .monitoring.rules.alerting set to true, the chart will fail to render.
```

**Comment by Daniel-Vaz** [comment]
SO can someone merge this pls ?

**Comment by dannykopping** [comment]
> [CLA assistant check](https://cla-assistant.io/grafana/loki?pullRequest=11241) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/grafana/loki?pullRequest=11241) before we can accept your contribution.You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/grafana/loki?pullRequest=11241) it.

@Daniel-Vaz please sign the CLA, and then I'll merge once CI completes its checks

**Comment by Daniel-Vaz** [comment]
@dannykopping Done, thank you 😄

**Comment by dannykopping** [comment]
@Daniel-Vaz there seem to be some CI failures (the [helm linter](https://github.com/grafana/loki/actions/runs/7246428746/job/19738369292?pr=11241) and the [helm docs](https://drone.grafana.net/grafana/loki/32355/7/2) checks, specifically).

Please address these, and then we can get this merged.

**Comment by Daniel-Vaz** [comment]
I updated the version reference to be `5.41.6`.
I ran helm-docs without issues.
But based on the drone ci workflow error I tried to run the command `make -BC docs sources/setup/install/helm/reference.md`, but I'm getting the following error:

```
$ make -BC docs sources/setup/install/helm/reference.md
make: Entering directory '/home/user/repos/loki/docs'
podman run --rm --volume "/home/user/repos/loki:/helm-docs" -u "$(id -u)" "docker.io/jnorwood/helm-docs:v1.11.0" \
        -c /helm-docs/production/helm/ \
        -t reference.md.gotmpl \
        -o reference.md
mv "../production/helm/loki/reference.md" "sources/setup/install/helm/reference.md"
time="2024-01-10T09:33:13Z" level=info msg="Found Chart directories [loki]"
time="2024-01-10T09:33:13Z" level=info msg="Generating README Documentation for chart /helm-docs/production/helm/loki"
time="2024-01-10T09:33:13Z" level=warning msg="Could not open chart README file /helm-docs/production/helm/loki/README.md, skipping chart"
mv: cannot stat '../production/helm/loki/reference.md': No such file or directory
make: *** [Makefile:15: sources/setup/install/helm/reference.md] Error 1
make: Leaving directory '/home/user/repos/loki/docs'
```

**Comment by dannykopping** [comment]
> I updated the version reference to be `5.41.6`. I ran helm-docs without issues. But based on the drone ci workflow error I tried to run the command `make -BC docs sources/setup/install/helm/reference.md`, but I'm getting the following error:
> 
> ```
> $ make -BC docs sources/setup/install/helm/reference.md
> make: Entering directory '/home/user/repos/loki/docs'
> podman run --rm --volume "/home/user/repos/loki:/helm-docs" -u "$(id -u)" "docker.io/jnorwood/helm-docs:v1.11.0" \
>         -c /helm-docs/production/helm/ \
>         -t reference.md.gotmpl \
>         -o reference.md
> mv "../production/helm/loki/reference.md" "sources/setup/install/helm/reference.md"
> time="2024-01-10T09:33:13Z" level=info msg="Found Chart directories [loki]"
> time="2024-01-10T09:33:13Z" level=info msg="Generating README Documentation for chart /helm-docs/production/helm/loki"
> time="2024-01-10T09:33:13Z" level=warning msg="Could not open chart README file /helm-docs/production/helm/loki/README.md, skipping chart"
> mv: cannot stat '../production/helm/loki/reference.md': No such file or directory
> make: *** [Makefile:15: sources/setup/install/helm/reference.md] Error 1
> make: Leaving directory '/home/user/repos/loki/docs'
> ```

Hhmm, i just ran it and it worked fine; did you run this from the root of the project?

You can always copy the changes in manually from this commit:
https://github.com/grafana/loki/commit/377c7f84c6c21a55375ec7aa4d7fd0185c06fe1f

**Comment by Daniel-Vaz** [comment]
@dannykopping 
I thinks its ready to be merged now.
Thanks

**Comment by dannykopping** [comment]
Oh man what a nightmare, this CI is failing on whitespace now 💀 🙃 
Thank you for your patience here @Daniel-Vaz! I think we're very nearly there.

**Comment by Daniel-Vaz** [comment]
@dannykopping 
No worries 😅  I'm, the one who is sorry for this mess.

I see the workflow failing indeed on what it seems to be a removed line\whitespace....
But on my side running `helm-docs` and `helm-lint` dont result in any change:

```sh
$ helm-docs
INFO[2024-01-17T07:53:09Z] Found Chart directories [production/helm/loki]
INFO[2024-01-17T07:53:09Z] Generating README Documentation for chart production/helm/loki

$ make helm-lint
make -BC production/helm/loki lint
make[1]: Entering directory '/home/user/repos/loki/production/helm/loki'
yamllint -c /home/user/repos/loki/production/helm/loki/src/.yamllint.yaml /home/user/repos/loki/production/helm/loki/src
make[1]: Leaving directory '/home/user/repos/loki/production/helm/loki'

$ git status
On branch main
Your branch is up to date with 'origin/main'.

nothing to commit, working tree clean
```

**Comment by dannykopping** [comment]
> @dannykopping No worries 😅 I'm, the one who is sorry for this mess.
> 
> I see the workflow failing indeed on what it seems to be a removed line\whitespace.... But on my side running `helm-docs` and `helm-lint` dont result in any change:
> 
> ```shell
> $ helm-docs
> INFO[2024-01-17T07:53:09Z] Found Chart directories [production/helm/loki]
> INFO[2024-01-17T07:53:09Z] Generating README Documentation for chart production/helm/loki
> 
> $ make helm-lint
> make -BC production/helm/loki lint
> make[1]: Entering directory '/home/user/repos/loki/production/helm/loki'
> yamllint -c /home/user/repos/loki/production/helm/loki/src/.yamllint.yaml /home/user/repos/loki/production/helm/loki/src
> make[1]: Leaving directory '/home/user/repos/loki/production/helm/loki'
> 
> $ git status
> On branch main
> Your branch is up to date with 'origin/main'.
> 
> nothing to commit, working tree clean
> ```

I think just manually delete that line then, and let's get this merged.
No need to keep merging `main` in, by the way.

**Comment by Daniel-Vaz** [comment]
Fingers crossed 🤞

**Comment by dannykopping** [comment]
😌 thanks for your patience @Daniel-Vaz! We appreciate the contribution a lot!

**Pull request merged**
